### ISSUE TITLE
Implement SHA3 instruction

### ISSUE THREAD

**Issue by elmattic** [elmattic/navi#23]

**Comment by elmattic** [comment]
First I want to thanks Markku-Juhani O. Saarinen for its small and readable implementation of Keccak.
Now let's talk about speed.

What is absolutely needed for top performances is that the compiler folds loads into instructions for the rho and pi stages.

Then I've found that `clang` would use vector instructions in `keccakf1600` and that was hurting performances A LOT. I don't understand why but ONLY `-fno-builtin` can disable that. Luckily it's a compiler option that I'm always using in my `CCFLAGS`, but that wasn't the case at first.
I will investigate on this.

At that point I was on par with the go client (which uses hand-rolled assembly on amd64 target).
So I've started to look at `BMI` and `BMI2` instructions for the EVM `AVX2` path.
Using them resulted in a 25% speed bump.

Here's the instruction breakdown with and without:

```
138 mov
 76 xor
 29 rorx
 25 andn
```

```
183 mov
 76 xor
 29 rol
 25 and
 25 not
```
